ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,294 @@
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment handlesVar
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
+            chainSyncStateVar <- cschState . (Map.! peer) <$> readTVar handlesVar
+            modifyTVar chainSyncStateVar $ \chainSyncState -> chainSyncState {csCandidate = slicedDynamoFragment}
+            setJumpingState handlesVar peer $ Jumper nextJumpVar goodPoint' Happy
+          (Happy, RejectedJump badPoint) ->
+            -- The previously happy peer just rejected a jump. We look for an
+            -- intersection.
+            lookForIntersection nextJumpVar goodPoint badPoint
+          (LookingForIntersection badPoint, AcceptedJump goodPoint') ->
+            -- The peer that was looking for an intersection accepts the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint' badPoint
+          (LookingForIntersection _, RejectedJump badPoint') ->
+            -- The peer that was looking for an intersection rejects the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint badPoint'
+  where
+    lookForIntersection nextJumpVar goodPoint badPoint = do
+      (dynamoFragment0 :: AnchoredFragment (Header blk)) <- fromJust <$> getDynamoFragment handlesVar -- full fragment
+      let dynamoFragment = fromJust $ sliceRangeExcl dynamoFragment0 goodPoint badPoint
+      case (toOldestFirst dynamoFragment :: [Header blk]) of
+        [] -> do
+          setJumpingState handlesVar peer $ Jumper nextJumpVar goodPoint FoundIntersection
+          electNewObjector handlesVar
+        points -> do
+          let (middlePoint :: Header blk) = points !! (length points `div` 2)
+          writeTVar nextJumpVar $ Just $! (castPoint (blockPoint middlePoint) :: Point blk)
+          setJumpingState handlesVar peer $ Jumper nextJumpVar goodPoint (LookingForIntersection badPoint)
+
+    -- | Select a slice of an anchored fragment between two points, inclusive.
+    -- Both points must exist on the chain, in order, or the result is Nothing.
+    -- REVIEW: How does this function behave if @fromExcl == toExcl@? How should
+    -- it behave? It doesn't matter much because we never have this in our code,
+    -- but still.
+    sliceRangeExcl fragment fromExcl toExcl = do
+      fragmentFrom <- snd <$> splitAfterPoint fragment fromExcl
+      fst <$> splitBeforePoint fragmentFrom toExcl
```

Comment:
```suggestion
    -- | Select a slice of an anchored fragment between two points. The slice
    -- includes neither of the points.
    --
    -- Both points must exist on the anchored fragment, in order,
    -- and `fromExcl /= toExcl`, otherwise the result is `Nothing`.
    sliceRangeExcl fragment fromExcl toExcl = do
      fragmentFrom <- snd <$> splitAfterPoint fragment fromExcl
      fst <$> splitBeforePoint fragmentFrom toExcl
```